### DISCUSSION TITLE
Public API for prop-fetching from SSR pages

### DISCUSSION THREAD

**Discussion by Schniz** [vercel/next.js#13714] [Ideas]
Hi there!

Imagine having the following page:

```js
// /pages/index.js

export default ({ date }) => {
  return (
    <div>
      Date: {date}.
      <Link href="/">
        <a>Refresh!</a>
      </Link>
    </div>
  );
}

export const getServerSideProps = () => {
  return {
    props: {
      date: new Date().getISOString()
    }
  }
}
```

When going to `/` we'll get `Hello, foo` and a link to `go to bar`.
When clicking on `go to bar`, Next.js issues a request to fetch the server side:
* In development, it'll be `/_next/data/development/index.json`
* In production it will be something like `/_next/data/<buildId>/index.json`

I believe that making this API public provides an opportunity for a wonderful DX with some use cases. Imagine you build a form and want to refresh your page afterwards with the new data. Instead of building a new API endpoint, you can ask the contents of this exact page. How would you do that? Probably, using the `Accept` header. When providing `text/html` in Accept, Next.js can render the entire page. When having `application/json` in the Accept header, Next can return the needed props for rendering the page.

This will work somewhat like [Intertia.js](https://inertiajs.com/)

What do you think?

**Comment by Schniz** [comment]
This can be done in userland using https://github.com/smeijer/next-runtime 😃

**Reply by smeijer** [reply]
Great minds think alike! This topic sounds indeed exactly like what `next-runtime` does. I don't do anything with the mentioned `/_next` routes tho, that's being left up to Next. I simply make it more convenient to handle `POST` requests (soon also the other non-get), and based on the `accept` header, I unwrap the returned props (`return acceptJson ? result.props : result`)

So there are indeed no changes required in Next to accomplish this. See for example [this sandbox](https://codesandbox.io/s/next-runtime-i50f3?file=/pages/index.js)
